proofs/pfedit.mli

```diff
@@ -190,4 +190,4 @@ val declare_implicit_tactic : unit Proofview.tactic -> unit
 val clear_implicit_tactic : unit -> unit
 
 (* Raise Exit if cannot solve *)
-val solve_by_implicit_tactic : env -> Evd.evar_map -> Evd.evar -> Evd.evar_map * constr
+val solve_by_implicit_tactic : unit -> (env -> Evd.evar_map -> Evd.evar -> Evd.evar_map * constr) option
```

Comment:
The inner function does raise Exit when it cannot solve the problem. I'm not sure I am willing to document this function a lot, as I would rather see it die at some point. The implementation is still flawed despite dae240f3, and I am not sure that people use the Solve Implicit stuff now that there is tactics-in-term. Furthermore, it ought to be in the monad at the very least.

The time saving comes from this piece of code in `Pretyping`:
```ocaml
  if Option.has_some flags.use_hook then
    apply_inference_hook (Option.get flags.use_hook env) evdref frozen;
```
where `apply_inference_hook` can be costly even if the tactic argument fails immediately, as it folds over all undefined evars:
```ocaml
  evdref := Evar.Set.fold (fun evk sigma ->
    if Evd.is_undefined sigma evk (* in particular not defined by side-effect *)
    then
      try
        let sigma, c = hook sigma evk in
        Evd.define evk c sigma
      with Exit ->
        sigma
    else
      sigma) pending !evdref
```